File: prolog/rules/use.pl
:- ensure_loaded([
    '../map/usable_items',
    'inventory',
    '../map/subplaces'
]).


use(Item, Subplace) :-
    holding(Item),
    current_pos(CurrentPlace),
    subplace(CurrentPlace, Subplace),
    on_use(Item, Subplace), !, nl.

use(small_key) :-
    holding(small_key),
    on_use(small_key, handcuffs),nl,!.

use(Item) :-
    holding(Item),
    current_pos(CurrentPlace),
    on_use(Item, CurrentPlace).

File: haskell/src/Feature/Use.hs
module Feature.Use where

import State.GameState (GameState(..), currentPlace)
import Object.Item (Item)
import Object.Place (Place)
import MapEventHandler.OnUse (onUseItem, onUsePlace, usableItemsPlace, usableItemsItem)
import Feature.Inventory (isInInventory)
import Feature.Subplace (subplaceExists)

useOnPlace :: Item -> Place -> GameState -> IO GameState
useOnPlace item place state = 
    if item `elem` usableItemsPlace place && (subplaceExists (currentPlace state) place state || place == currentPlace state) && isInInventory item state
        then onUsePlace item place state
        else do
            putStrLn "You can't use that item on that place!"
            return state

useOnItem :: Item -> Item -> GameState -> IO GameState
useOnItem item1 item2 state =
    if item1 `elem` usableItemsItem item2 && isInInventory item1 state && isInInventory item2 state then
        onUseItem item1 item2 state
    else do
            putStrLn "You can't use that item on that item!"
            return state